### ISSUE TITLE
Rebuild dependencies from scratch adding a compiler flag?

### ISSUE THREAD

**Issue by hamstergene** [conan-io/conan#2417]
Is there a way to tell Conan to rebuild all dependencies with the same compiler flag? The kind of flag that is applied to every *.c/*.cpp file, both in my project and every statically linked dependency (e.g. enable address sanitizer, or if I want to force -Os on everything).

This is strictly for one project, I do not want unrelated projects to accidentally pick up those products from the cache.

Can this be done without manually duplicating, editing and re-exporting recipes of every dependency?

**Comment by memsharded** [comment]
As conan packages wrap the underlying build systems, there is no general way to do it without such build systems be collaborative. 

In the OSS case, many existing and popular OSS packages have build systems that are not compatible with main practices, e.g. OpenSSL. No matter what environment variables you define, the OpenSSL build will not use them. There is nothing you can do, except explicitly telling from the conanfile.py recipe to use such special flags. That needs in many cases even patching the build scripts. So most existing OSS packages do not provision ways to inject to the underlying build system any arbitrary flag or configuration. Some of them will provide some package options to configure things, but outside of those options, you can't do anything.

There are OSS packages which build systems are "standard", for example in the case of using Autotools. In that case you can well define environment variables as CXXFLAGS, and they will be used in the build of the package. Note however, that this is not modeled in the conan binary package ID. If you want to properly have conan packages for example for sanitized/non-sanitized, you need to define a package option that will model that different binary. Everything that affects and produces a different final binary should be encoded as settings or options, to let conan know about it and manage binaries properly.

So to summarize no, you can't do it. For your own recipes, it is easy to provision them with the model (settings/options) you want to model that, for existing third-party packages, there is no way. This is an unfortunate consequence of build systems not providing some programming uniform interface (or at least an interface) to define such things.

**Comment by hamstergene** [comment]
> If you want to properly have conan packages for example for sanitized/non-sanitized, you need to define a package option that will model that different binary.

These flags have to be the same for everything, because it does not make sense or will not work to set them separately for each package. They are configuration/toolchain level, they can not be package-level options.

There are `-mmacos-version-min`, `-mios-version-min` which may make prebuilt binaries wrong/unusable even when compiler/arch/etc are exactly the same. Or one may want to use modified/nightly version of clang with some diagnostic/codegen flag that does not exist yet. Bottom line is that recipes do not even need to analyze them, they just need to pass them to the build system. It's like different compiler version, except you wouldn't want to hardcode it into `settings.yml`.

Maybe there should be universal profile settings like `compiler.cflags` and `compiler.cppflags`? Such that changing them will just assume different binary IDs.

**Comment by memsharded** [comment]
Yes, we have considered such universal settings, but they have a few problems:
- They don't keep the order. Providing ``-Wall -Werror`` would produce a different binary than ``-Werror -Wall``. Doing an ordering of the flags, I don't think that is safe.
- They don't necesarily affect or produce a different binary. Adding flags ``-Wall`` won't produce a different binary than if not provided. But that will exponentially increase the number of binaries (and also the chances that there will not be a pre-built binary for your configuration)
- Even if you have them, it is extremely difficult to map them in a general way to all build systems. What about Windows? What to do if the first thing you have in an OSS library CMakeLists is an override of CMAKE_CXX_FLAGS? It is not possible to do this, until build systems are really prepared for it.

I am sorry, this is something that I'd love myself to be able to have. But being realistic, it is not something that any tool that integrates with any build system can provide seamlessly. So the conan approach is to be pragmatic and provide the tools so users can customize things as they want, not trying to be "one size fits all" solution.

The right approach that we have already discussed is to explicitly model that. As sanitizing is becoming mainstream, it could make sense to define ``sanitizer`` settings, in the same way you have ``runtime`` or ``libcxx``. This is not in our inmediate roadmap, as users that want to do that can already customize the packages, and the wider OSS community is not that concerned about it.

**Comment by hamstergene** [comment]
Well I kinda have achieved it: added those settings to `settings.yml` and set them to same values as  `CFLAGS`/`CXXFLAGS` env vars. Now whenever the flags change a separate set of binaries gets built, separate from other projects. Most build systems pick up environment automatically so it works for everything except OpenSSL. That means just one recipe to customize instead of dozens, I'll take it.

I of course do not propose replacing current settings with arbitrary flags. That would be bad even if it worked. It's more about having an advanced "I know what I'm doing" option for non-standard cases. Having standard documented source of extra flags would offload figuring how to pass them to the recipe authors.

**Comment by memsharded** [comment]
There is a new way to inject flags in new build system integrations, via ``tools.build:cxxflags`` ``conf``, and there is even a way that we can make the ``conf`` affect the ``package_id`` in Conan 2.0, so this would implement this flow.

Closing this as inactive.
